api/rtx/inc/uvisor_allocator.h

```
@@ -0,0 +1,83 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef __UVISOR_RTX_ALLOCATOR_H__
+#define __UVISOR_RTX_ALLOCATOR_H__
+
+#include <stdint.h>
+#include <stddef.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** Contains the allocator data and backing page table. */
+typedef void * UvisorAllocator;
+
+/** Create an allocator in-place in an existing pool without using pages.
+ * Use this to turn statically allocated memory into a heap.
+ * Or allocate a large piece of memory and then turn that into a heap.
+ *
+ * @param mem   Pointer to the origin of the memory pool
+ * @param bytes Length of the memory pool in bytes
+ * @returns the allocator or `NULL` on failure
+ */
+UvisorAllocator uvisor_allocator_create_with_pool(
+    void * mem,
+    size_t bytes);
+
+/** Create an allocator using pages from the page heap.
+ * Use this to request secure dynamic memory for your process.
+ * Note that this memory is not guaranteed to be consequtive, therefore you
+ * must specify the maximum allocation size that you plan to use in this
+ * allocator. This function will then compute the number and size of required
+ * pages and request them from the secure page heap.
+ *
+ * @param total_size          The minimal total size of the heap
+ * @param maximum_malloc_size The largest size to be allocated in one chunk
+ * @returns the allocator or `NULL` on failure (out of memory,
+ *          maximum malloc size cannot be fulfilled)
+ */
+UvisorAllocator uvisor_allocator_create_with_pages(
+    size_t total_size,
+    size_t maximum_malloc_size);
+
+/** Destroy the allocator and free the backing pages.
+ * You cannot destroy a memory-pool backed allocator! */
```

Does it fail explicitly, silently or via a return value?